var/ramble/repos/builtin/package_managers/pip/package_manager.py

```diff
@@ -0,0 +1,475 @@
+        self.installed = False
+
+    def configure_env(self, path):
+        """Configure the venv path for subsequent commands"""
+        self.env_path = path
+
+    def set_dry_run(self, dry_run=False):
+        """Set the dry_run state of this pip runner"""
+        self.dry_run = dry_run
+
+    def create_env(self, env_path):
+        """Ensure a venv environment is created"""
+        if os.path.exists(env_path) and not os.path.isdir(env_path):
+            raise RunnerError(f"Unable to create environment {env_path}")
+
+        if not os.path.exists(env_path):
+            fs.mkdirp(env_path)
+
+        if not self.dry_run:
+            if not os.path.exists(os.path.join(env_path, self._venv_name)):
+                self.bs_python(
+                    "-m", "venv", os.path.join(env_path, self._venv_name)
+                )
+
+        # Ensure subsequent commands use the created env now.
+        self.env_path = env_path
+
+    def _get_venv_python(self):
+        if self.dry_run:
+            return self.bs_python.copy()
+        return Executable(
+            os.path.join(self.env_path, self._venv_name, "bin", "python")
+        )
+
+    def install(self):
+        """Invoke pip install"""
+        self._check_env_configured()
+        if self.installed:
+            logger.debug("Installation already done, skipping")
+            return
+        req_file = os.path.join(self.env_path, self._requirement_file_name)
+        if not os.path.exists(req_file):
+            raise RunnerError(f"{req_file} does not exist")
+        installer = self._get_venv_python()
+        installer.add_default_arg("-m")
+        installer.add_default_arg("pip")
+        installer_flags = ramble.config.get(
+            f"{self.install_config_name}:flags"
+        )
+        install_args = ["install", "-r", req_file, *installer_flags]
+        freeze_args = ["freeze", "-r", req_file]
+        if self.dry_run:
+            self._dry_run_print(installer, install_args)
+            self._dry_run_print(installer, freeze_args)
+        else:
+            installer(*install_args)
+            lock_file = os.path.join(self.env_path, self._lock_file_name)
+            with open(lock_file, "w") as f:
+                installer(*freeze_args, output=f)
+        self.installed = True
+
+    def _get_activate_script_path(self):
+        return os.path.join(self.env_path, self._venv_name, "bin", "activate")
+
+    def _check_env_configured(self):
+        """Check if virtual env is configured"""
+        if not self.env_path:
+            raise RunnerError("env_path is not configured")
+        if self.dry_run:
+            return
+        script_path = self._get_activate_script_path()
+        if not os.path.exists(script_path):
+            raise RunnerError("virtual env is not configured")
+
+    def generate_activate_command(self):
+        """Generate a command to activate a virtual env"""
+        return [f"source {self._get_activate_script_path()}"]
```

Comment:
We probably want to use `.` instead of `source`. You could even use https://github.com/douglasjacobsen/ramble/blob/package_manager_abstraction/lib/ramble/ramble/util/sourcing.py to get the source command for the correct shell.

Reply:
Good call, updated.